src/TurbulenceClosures/turbulence_closure_implementations/CATKEVerticalDiffusivities/CATKEVerticalDiffusivities.jl

```diff
@@ -249,47 +254,76 @@ function compute_diffusivities!(diffusivities, closure::FlavorOfCATKE, model; pa
     Δt = model.clock.time - diffusivities.previous_compute_time[]
     diffusivities.previous_compute_time[] = model.clock.time
 
+    launch!(arch, grid, parameters,
+            _compute_CATKE_auxiliaries!,
+            diffusivities, grid, closure, velocities, tracers, buoyancy)
+
     launch!(arch, grid, :xy,
-            compute_average_surface_buoyancy_flux!,
-            diffusivities.Qᵇ, grid, closure, velocities, tracers, buoyancy, top_tracer_bcs, clock, Δt)
+            _compute_average_surface_buoyancy_flux!,
+            diffusivities.Qᵇ, grid, closure, diffusivities, velocities, tracers, buoyancy, top_tracer_bcs, clock, Δt)
 
     launch!(arch, grid, parameters,
-            compute_CATKE_diffusivities!,
+            _compute_CATKE_diffusivities!,
             diffusivities, grid, closure, velocities, tracers, buoyancy)
 
     return nothing
 end
 
-@kernel function compute_average_surface_buoyancy_flux!(Qᵇ, grid, closure, velocities, tracers, buoyancy, top_tracer_bcs, clock, Δt)
+@kernel function _compute_CATKE_auxiliaries!(diffusivities, grid, closure, velocities, tracers, buoyancy)
+    i, j, k = @index(Global, NTuple)
+
+    e  = tracers.e
+    S² = diffusivities.S²
+    N² = diffusivities.N²
+    w★ = diffusivities.w★
+    u, v, _ = velocities
+
+    @inbounds begin
+        S²[i, j, k] = shearᶜᶜᶠ(i, j, k, grid, u, v)
+        N²[i, j, k] = ∂z_b(i, j, k, grid, buoyancy, tracers)
+        w★[i, j, k] = turbulent_velocityᶜᶜᶜ(i, j, k, grid, closure, tracers.e)
```

Comment:
Does this really need to be precomputed? It's just the sqrt of `e`.

Reply:
It seems like the big one here is precomputing the buoyancy gradient, since that's potentially expensive.

Rather than completely nuking the existing code, it might make sense to generalize it instead so we can avoid this memory allocation if desired.

Reply:
We probably don't even need that extra memory. It can be precomputed inside the `_compute_CATKE_diffusivities!` kernel (at centers and faces).